File: python/react-custom-backend/src/app/threads.py
"""Thread state helpers backed by the graph checkpointer.

Implements the LangGraph SDK thread state wire-shape consumed by
``client.threads.getState`` / ``updateState`` (``GET|POST /threads/:id/state``),
aligned with the Agent Protocol thread model in ``agent-protocol/openapi.json``.
"""

from __future__ import annotations

from typing import Any

from langchain_core.messages import BaseMessage
from langgraph.graph.state import CompiledStateGraph
from langgraph.types import StateSnapshot

LocalProtocolGraph = CompiledStateGraph

# Empty thread bootstrap must write through ``__start__`` so conditional edges
# are not evaluated on an empty ``messages`` list. Non-empty updates on an
# existing checkpoint use ``agent`` when the client omits ``as_node``.
INITIAL_UPDATE_NODE = "__start__"
DEFAULT_UPDATE_NODE = "agent"


def thread_config(thread_id: str) -> dict[str, Any]:
    return {"configurable": {"thread_id": thread_id}}


def thread_has_checkpoint(snapshot: StateSnapshot) -> bool:
    configurable = snapshot.config.get("configurable", {})
    checkpoint_id = configurable.get("checkpoint_id")
    return isinstance(checkpoint_id, str) and bool(checkpoint_id)


def _serialize_value(value: Any) -> Any:
    if isinstance(value, BaseMessage):
        return value.model_dump()
    if isinstance(value, list):
        return [_serialize_value(item) for item in value]
    if isinstance(value, dict):
        return {key: _serialize_value(item) for key, item in value.items()}
    return value


def serialize_thread_state(snapshot: StateSnapshot, thread_id: str) -> dict[str, Any]:
    """Serialize a LangGraph ``StateSnapshot`` to the SDK ``ThreadState`` shape."""
    configurable = snapshot.config.get("configurable", {})
    checkpoint_id = configurable.get("checkpoint_id")
    checkpoint_ns = configurable.get("checkpoint_ns", "")

    tasks: list[dict[str, Any]] = []
    for task in snapshot.tasks or ():
        if not isinstance(task, dict):
            continue
        tasks.append(
            {
                "id": task.get("id"),
                "name": task.get("name"),
                "error": task.get("error"),
                "interrupts": task.get("interrupts") or [],
                "checkpoint": task.get("checkpoint"),
                "state": task.get("state"),
            }
        )

    return {
        "values": _serialize_value(snapshot.values or {}),
        "next": list(snapshot.next or ()),
        "tasks": tasks,
        "checkpoint": {
            "thread_id": thread_id,
            "checkpoint_id": checkpoint_id,
            "checkpoint_ns": checkpoint_ns,
        },
        "metadata": dict(snapshot.metadata or {}),
        "created_at": None,
        "parent_checkpoint": None,
    }


async def get_thread_state(graph: LocalProtocolGraph, thread_id: str) -> dict[str, Any]:
    snapshot = await graph.aget_state(thread_config(thread_id))
    if not thread_has_checkpoint(snapshot):
        raise KeyError(thread_id)
    return serialize_thread_state(snapshot, thread_id)


def _parse_before_cursor(
    thread_id: str, before: Any
) -> dict[str, Any] | None:
    if before is None:
        return None
    if isinstance(before, str):
        return {
            "configurable": {
                "thread_id": thread_id,
                "checkpoint_id": before,
            }
        }
    if isinstance(before, dict):
        configurable = before.get("configurable")
        if not isinstance(configurable, dict):
            configurable = before
        checkpoint_id = configurable.get("checkpoint_id")
        if not isinstance(checkpoint_id, str):
            return None
        cursor: dict[str, Any] = {
            "configurable": {
                "thread_id": thread_id,
                "checkpoint_id": checkpoint_id,
            }
        }
        checkpoint_ns = configurable.get("checkpoint_ns")
        if isinstance(checkpoint_ns, str):
            cursor["configurable"]["checkpoint_ns"] = checkpoint_ns
        return cursor
    return None


async def get_thread_history(
    graph: LocalProtocolGraph,
    thread_id: str,
    *,
    limit: int = 10,
    before: Any = None,
) -> list[dict[str, Any]]:
    try:
        await get_thread_state(graph, thread_id)
    except KeyError:
        raise KeyError(thread_id) from None

    history: list[dict[str, Any]] = []
    before_config = _parse_before_cursor(thread_id, before)
    async for snapshot in graph.aget_state_history(
        thread_config(thread_id),
        before=before_config,
        limit=limit,
    ):
        history.append(serialize_thread_state(snapshot, thread_id))
    return history


def _resolve_update_node(
    *,
    as_node: str | None,
    values: dict[str, Any] | None,
    has_checkpoint: bool,
) -> str:
    if as_node:
        return as_node

    messages = (values or {}).get("messages")
    if not messages:
        return INITIAL_UPDATE_NODE
    if not has_checkpoint:
        return INITIAL_UPDATE_NODE
    return DEFAULT_UPDATE_NODE


async def update_thread_state(
    graph: LocalProtocolGraph,
    thread_id: str,
    *,
    values: dict[str, Any] | None = None,
    checkpoint: dict[str, Any] | None = None,
    as_node: str | None = None,
) -> dict[str, Any]:
    config = thread_config(thread_id)
    if checkpoint and isinstance(checkpoint.get("checkpoint_id"), str):
        config = {
            "configurable": {
                **config["configurable"],
                "checkpoint_id": checkpoint["checkpoint_id"],
                **(
                    {"checkpoint_ns": checkpoint["checkpoint_ns"]}
                    if isinstance(checkpoint.get("checkpoint_ns"), str)
                    else {}
                ),
            }
        }

    snapshot = await graph.aget_state(config)
    resolved_values = values if values is not None else {"messages": []}
    resolved_as_node = _resolve_update_node(
        as_node=as_node,
        values=resolved_values,
        has_checkpoint=thread_has_checkpoint(snapshot),
    )

    await graph.aupdate_state(
        config,
        resolved_values,
        as_node=resolved_as_node,
    )
    snapshot = await graph.aget_state(thread_config(thread_id))
    return serialize_thread_state(snapshot, thread_id)

File: typescript/react-custom-backend/src/server/threads.ts
/**
 * Thread state helpers backed by the graph checkpointer.
 *
 * Port of the Python example's `app/threads.py`. Implements the LangGraph SDK
 * thread state wire-shape consumed by `client.threads.getState` / `updateState`
 * (`GET|POST /threads/:id/state`) and `getHistory` (`POST /threads/:id/history`),
 * aligned with the Agent Protocol thread model.
 */

import type { CompiledGraphType } from "@langchain/langgraph";
import type { RunnableConfig } from "@langchain/core/runnables";

import { isRecord, sanitizeForJson } from "./serialize.js";

/**
 * Compiled LangGraph instance exposed through the custom protocol server.
 *
 * Thread routes read and write checkpointed state through this graph's
 * checkpointer rather than maintaining a separate thread store.
 */
export type LocalProtocolGraph = CompiledGraphType;

type StateSnapshot = Awaited<ReturnType<LocalProtocolGraph["getState"]>>;

/**
 * Raised when a thread has no checkpoint yet.
 *
 * {@link CustomServer} maps this to HTTP 404 so the LangGraph SDK can
 * bootstrap the thread via `POST /threads/:id/state` before the first run.
 */
export class ThreadNotFoundError extends Error {
  readonly threadId: string;

  /**
   * @param threadId - Thread id that has not been checkpointed yet.
   */
  constructor(threadId: string) {
    super(`Thread ${threadId} not found`);
    this.name = "ThreadNotFoundError";
    this.threadId = threadId;
  }
}

/**
 * Graph node used when bootstrapping an empty thread.
 *
 * Empty `messages` updates must land on `__start__` so conditional edges are
 * not evaluated before the first human turn exists.
 */
const INITIAL_UPDATE_NODE = "__start__";

/**
 * Default graph node for non-empty state updates on an existing checkpoint.
 *
 * Matches `createAgent`'s model node when the client omits `as_node`.
 */
const DEFAULT_UPDATE_NODE = "model_request";

/**
 * Build the {@link RunnableConfig} that scopes graph calls to a thread id.
 *
 * @param threadId - Stable conversation id from the Agent Protocol layer.
 * @returns Config whose `configurable.thread_id` selects the checkpointer row.
 */
function threadConfig(threadId: string): RunnableConfig {
  return { configurable: { thread_id: threadId } };
}

/**
 * Read the `configurable` bag from a LangGraph run config.
 *
 * @param config - Runnable config returned by `getState` / passed to updates.
 * @returns Plain object (empty when `configurable` is missing or not a record).
 */
function configurableOf(config: RunnableConfig): Record<string, unknown> {
  return isRecord(config.configurable) ? config.configurable : {};
}

/**
 * Return whether a {@link StateSnapshot} represents a persisted checkpoint.
 *
 * LangGraph returns an empty configurable bag before the first write; the SDK
 * treats that as "thread not found" rather than an empty thread state.
 *
 * @param snapshot - Snapshot from `graph.getState`.
 * @returns `true` when `checkpoint_id` is a non-empty string.
 */
function threadHasCheckpoint(snapshot: StateSnapshot): boolean {
  const checkpointId = configurableOf(snapshot.config).checkpoint_id;
  return typeof checkpointId === "string" && checkpointId.length > 0;
}

/**
 * Serialize a LangGraph {@link StateSnapshot} to the SDK `ThreadState` shape.
 *
 * Converts LangChain message instances in `values` to plain JSON, normalizes
 * pending `tasks`, and fills the `checkpoint` envelope expected by
 * `client.threads.getState` consumers.
 *
 * @param snapshot - Raw snapshot from `graph.getState` or `getStateHistory`.
 * @param threadId - Thread id echoed in the response `checkpoint.thread_id`.
 * @returns JSON-serializable thread state for HTTP responses.
 */
export function serializeThreadState(
  snapshot: StateSnapshot,
  threadId: string
): Record<string, unknown> {
  const configurable = configurableOf(snapshot.config);
  const checkpointId =
    typeof configurable.checkpoint_id === "string"
      ? configurable.checkpoint_id
      : null;
  const checkpointNs =
    typeof configurable.checkpoint_ns === "string"
      ? configurable.checkpoint_ns
      : "";

  const tasks = (snapshot.tasks ?? []).map((task) => {
    const record = task as {
      id?: unknown;
      name?: unknown;
      error?: unknown;
      interrupts?: unknown;
      state?: unknown;
    };
    return {
      id: record.id,
      name: record.name,
      error: record.error ?? null,
      interrupts: Array.isArray(record.interrupts) ? record.interrupts : [],
      state: record.state ?? null,
    };
  });

  return {
    values: sanitizeForJson(snapshot.values ?? {}),
    next: [...(snapshot.next ?? [])],
    tasks,
    checkpoint: {
      thread_id: threadId,
      checkpoint_id: checkpointId,
      checkpoint_ns: checkpointNs,
    },
    metadata: { ...snapshot.metadata },
    created_at: snapshot.createdAt ?? null,
    parent_checkpoint: null,
  };
}

/**
 * Read checkpointed thread state for `GET /threads/:threadId/state`.
 *
 * @param graph - Compiled agent graph with an attached checkpointer.
 * @param threadId - Thread id from the URL path.
 * @returns Serialized thread state in SDK wire format.
 * @throws {@link ThreadNotFoundError} When the thread has no checkpoint yet.
 */
export async function getThreadState(
  graph: LocalProtocolGraph,
  threadId: string
): Promise<Record<string, unknown>> {
  const snapshot = await graph.getState(threadConfig(threadId));
  if (!threadHasCheckpoint(snapshot)) throw new ThreadNotFoundError(threadId);
  return serializeThreadState(snapshot, threadId);
}

/**
 * Parse the `before` pagination cursor accepted by `POST /threads/:id/history`.
 *
 * The SDK may send either a checkpoint id string or a partial checkpoint
 * object from a prior `ThreadState.checkpoint` field.
 *
 * @param threadId - Thread id to scope the history query.
 * @param before - Opaque cursor from the request body (or `undefined`).
 * @returns Runnable config for `getStateHistory`, or `undefined` to start at head.
 */
function parseBeforeCursor(
  threadId: string,
  before: unknown
): RunnableConfig | undefined {
  if (before == null) return undefined;
  if (typeof before === "string") {
    return { configurable: { thread_id: threadId, checkpoint_id: before } };
  }
  if (!isRecord(before)) return undefined;

  const configurable = isRecord(before.configurable) ? before.configurable : before;
  const checkpointId = configurable.checkpoint_id;
  if (typeof checkpointId !== "string") return undefined;

  const cursor: RunnableConfig = {
    configurable: { thread_id: threadId, checkpoint_id: checkpointId },
  };
  if (typeof configurable.checkpoint_ns === "string") {
    (cursor.configurable as Record<string, unknown>).checkpoint_ns =
      configurable.checkpoint_ns;
  }
  return cursor;
}

/**
 * List past thread states for `POST /threads/:threadId/history`.
 *
 * Verifies the thread exists (404 when missing) before walking
 * `graph.getStateHistory`, matching {@link getThreadState} behavior.
 *
 * @param graph - Compiled agent graph with an attached checkpointer.
 * @param threadId - Thread id from the URL path.
 * @param options.limit - Maximum snapshots to return (default `10`).
 * @param options.before - Checkpoint cursor; omit to read from the latest state.
 * @returns Newest-first list of serialized thread states.
 * @throws {@link ThreadNotFoundError} When the thread has no checkpoint yet.
 */
export async function getThreadHistory(
  graph: LocalProtocolGraph,
  threadId: string,
  options: { limit?: number; before?: unknown } = {}
): Promise<Record<string, unknown>[]> {
  await getThreadState(graph, threadId);

  const history: Record<string, unknown>[] = [];
  const iterator = graph.getStateHistory(threadConfig(threadId), {
    before: parseBeforeCursor(threadId, options.before),
    limit: options.limit ?? 10,
  });
  for await (const snapshot of iterator) {
    history.push(serializeThreadState(snapshot, threadId));
  }
  return history;
}

/**
 * Choose which graph node should receive a `updateState` write.
 *
 * @param options.asNode - Explicit node from the client (`as_node` body field).
 * @param options.values - State payload; empty `messages` triggers bootstrap path.
 * @param options.hasCheckpoint - Whether the thread already has a checkpoint row.
 * @returns Node name passed to `graph.updateState` as the update attribution target.
 */
function resolveUpdateNode(options: {
  asNode?: string;
  values: Record<string, unknown> | null;
  hasCheckpoint: boolean;
}): string {
  if (options.asNode) return options.asNode;
  const messages = options.values?.messages;
  if (!Array.isArray(messages) || messages.length === 0) {
    return INITIAL_UPDATE_NODE;
  }
  if (!options.hasCheckpoint) return INITIAL_UPDATE_NODE;
  return DEFAULT_UPDATE_NODE;
}

/**
 * Create or update thread state for `POST /threads/:threadId/state`.
 *
 * Used by the browser bootstrap in `src/app/threads.ts` and by the SDK when
 * hydrating or editing conversation history. Applies the update at
 * {@link resolveUpdateNode}, then returns the latest serialized snapshot.
 *
 * @param graph - Compiled agent graph with an attached checkpointer.
 * @param threadId - Thread id from the URL path.
 * @param options.values - State patch (defaults to `{ messages: [] }` for create).
 * @param options.checkpoint - Optional fork point (`checkpoint_id`, `checkpoint_ns`).
 * @param options.asNode - Optional explicit update node (`as_node` body field).
 * @returns Serialized thread state after the write.
 */
export async function updateThreadState(
  graph: LocalProtocolGraph,
  threadId: string,
  options: {
    values?: Record<string, unknown> | null;
    checkpoint?: Record<string, unknown> | null;
    asNode?: string;
  } = {}
): Promise<Record<string, unknown>> {
  let config = threadConfig(threadId);
  const checkpoint = options.checkpoint;
  if (checkpoint && typeof checkpoint.checkpoint_id === "string") {
    config = {
      configurable: {
        ...configurableOf(config),
        checkpoint_id: checkpoint.checkpoint_id,
        ...(typeof checkpoint.checkpoint_ns === "string"
          ? { checkpoint_ns: checkpoint.checkpoint_ns }
          : {}),
      },
    };
  }

  const snapshot = await graph.getState(config);
  const resolvedValues = options.values ?? { messages: [] };
  const resolvedAsNode = resolveUpdateNode({
    asNode: options.asNode,
    values: resolvedValues,
    hasCheckpoint: threadHasCheckpoint(snapshot),
  });

  await graph.updateState(config, resolvedValues, resolvedAsNode);
  const updated = await graph.getState(threadConfig(threadId));
  return serializeThreadState(updated, threadId);
}
